src/analysis/simulation.jl

```diff
@@ -7,172 +7,202 @@ wm_solution.
-        minor_loss = wn.links._data[cv_name].minor_loss
-        status = wn.links._data[cv_name].status
+
+
+    ## add short_pipes
+    for (short_pipe_id,short_pipe_info) in wm_data["nw"]["1"]["short_pipe"]
+        node_fr = string(short_pipe_info["node_fr"])
+        node_to = string(short_pipe_info["node_to"])
+        length = 1e-3
+        diameter = 1
+        roughness = 100
+        minor_loss = short_pipe_info["minor_loss"]
+        status = short_pipe_info["status"]
         if status == 1
             status = "Open"
         else
             status = "Closed"
         end
-        cv_flag = false # now we already have a new zero-length CV, this pipe doesn't have to be cv again
+        wn.add_pipe("short_pipe"*short_pipe_id,node_fr,node_to,length=length,diameter=diameter,
+                    roughness=roughness,minor_loss=minor_loss,status=status)
+    end
+
+
+    ## add valves
+    for (valve_id,valve_info) in wm_data["nw"]["1"]["valve"]
+        node_fr = string(valve_info["node_fr"])
+        node_to = string(valve_info["node_to"])
+        length = 1e-3
+        diameter = 1
+        roughness = 100
+        minor_loss = 0
+        status = "Open"
+        if string(valve_info["flow_direction"]) == "UNKNOWN"
+            cv_flag = false
+        elseif string(valve_info["flow_direction"]) == "POSITIVE"
+            cv_flag = true
+        end
+        wn.add_pipe("valve"*valve_id,node_fr,node_to,length=length,diameter=diameter,
+                    roughness=roughness,minor_loss=minor_loss,status=status,check_valve_flag = cv_flag)
+    end
+
 
-        # remove link
-        wn.remove_link(cv_name,with_control=true,force=true)
-        # add the link back and connect to the artificial node
-        wn.add_pipe(cv_name,start_node_name=arti_cv_node_dict[cv_name],end_node_name=end_node_name,
-            length=cv_length,diameter=diameter,roughness=roughness,minor_loss=minor_loss,status=status,check_valve_flag=cv_flag)
+    ## add regulators
+    # TODO
+
+    ## add pumps
+    for (pump_id,pump_info) in wm_data["nw"]["1"]["pump"]
+        # add pump with attributes
+        node_fr = string(pump_info["node_fr"])
+        node_to = string(pump_info["node_to"])
+        head_curve = pump_info["head_curve"]
+        wn.add_curve("pump_head_curve"*pump_id,"HEAD",head_curve)
+        wn.add_pump("pump"*pump_id,node_fr,node_to,pump_type="HEAD",pump_parameter="pump_head_curve"*pump_id,
+                    speed=1,pattern=nothing)
+
+        # couple efficiency_curve to the pump
+        if haskey(pump_info,"efficiency_curve")
+            efficiency_curve = pump_info["efficiency_curve"]
+            wn.add_curve("pump_eff_curve"*pump_id,"EFFICIENCY",efficiency_curve)
+            wn.get_link("pump"*pump_id).efficiency = wn.get_curve("pump_eff_curve"*pump_id)
+        else
+            wn.options.energy.global_efficiency = pump_info["efficiency"]
+        end
     end
 
+
+
+
+
     # remove old controls
```

Comment:
If we are creating the WNTR network from scratch, I don't think there are any old controls to remove?

Reply:
@jjstickel Thank you for these comments, I will address them and make another commit.